sample-prep/Biomek_pooling_VI.py

```diff
@@ -29,107 +43,287 @@ def main(process_id, file_id):
+    # over or under threshold = red, within threshold = palegreen, adjusted volume to max = yellow 
+
+    volume_threshold_plotcolor = []
+
+    for volumes in sample_volumes:
+        if volumes > max_vol_available or volumes < min_vol_required:
+            volume_threshold_plotcolor.append("red")
+        else:
+            volume_threshold_plotcolor.append("palegreen") 
+
+    if enforce_max_vol == True:
+        for i in range(len(sample_volumes)):
+            if sample_volumes[i] > max_vol_available:
+                sample_volumes[i] = max_vol_available
+                volume_threshold_plotcolor[i] = "yellow"
 
-    #write csv-file
+    if enforce_min_volume == True:
+        if min_volume_in_list < min_vol_required:
+            volume_threshold_plotcolor[min_volume_in_list_index] = "yellow"
+
+    # DEFINE BIOMEK DECK PLATE POSITIONS (LIBLABWARE)
+    # create list with biomek LibLabware variable for sorting plate position on biomek deck 
+    # (plate 1 = position "Libraries 1", plate 2 = position "Libraries 2")
+    # no duplicate values allowed in set, so can extract unique container names by redefining to set. 
+
+    input_container_name_unique = sorted(list(set(input_container_name)))
+    print(input_container_name_unique)
+    libLabware = []
+    for i in range(len(input_container_name)):
+        if input_container_name[i] == input_container_name_unique[0]:
+            libLabware.append("Libraries1")
+        elif input_container_name[i] == input_container_name_unique[1]:
+            libLabware.append("Libraries2")
+        else:
+            print("Error assigning Biomek LibLabware positions. Contact system admin")
+            sys.exit(1)
+
+    #WRITE CSV FILE 
+    # using data loaded above. Looping over all inputs and writing line-by-line to the csv-file
+
+    total_sample_number = len(sample_name)
 
     with open(file_id + "-Biomek_pooling.csv", "w") as file:
-        file.write("LibLabware,LibWellPos,DestLabware,DestWellPos,LibVol,SampleName,Molarity,Species,GenomeSize,ProjectAccount" + '\n')
+        file.write("LibLabware,LibWellPos,DestLabware,DestWellPos,LibVol,SampleName,Molarity,Species,GenomeSize,PlateID,fmol/Mbp-actual" + '\n')
         for i in range(total_sample_number):
             file.write(
-                "Libraries" + ',' + 
+                str(libLabware[i]) + ',' + 
                 str(re.sub(r':','',source_location[i][1])) + ',' + 
                 'LibraryPool'  + ',' + 
                 "D1" + "," + 
-                str(sample_volumes[i]) + "," + 
+                str(round(sample_volumes[i],1)) + "," + 
                 str(sample_name[i]) + "," + 
                 str(molarity[i]) + "," +
                 str(species[i]) + "," + 
                 str(genome_size[i]) + "," +
-                str(project_account[i]) + "," + '\n')
+                str(input_container_name[i]) + "," +
+                str(round((sample_volumes[i]*molarity[i])/genome_size[i],0))+ '\n')
```

Comment:
You could consider interpolating this text too, I think you can do it with just adding a list of values after.

Reply:
I am not sure what you mean by interpolating the text? 

But this part is completely different now anyways. Now I am collecting all the data needed into a pandas dataframe and writing that dataframe to csv using the to_csv() function from pandas. Makes it much easier to read, I think. Also removed a lot of columns that I had there for testing/verification and not needed for produciton